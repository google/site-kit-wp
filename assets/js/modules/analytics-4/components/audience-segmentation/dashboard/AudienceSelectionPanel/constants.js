/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
