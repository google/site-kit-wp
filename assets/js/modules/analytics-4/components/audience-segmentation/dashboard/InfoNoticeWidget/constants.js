/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
