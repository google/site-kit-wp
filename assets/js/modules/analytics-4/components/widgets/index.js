/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
