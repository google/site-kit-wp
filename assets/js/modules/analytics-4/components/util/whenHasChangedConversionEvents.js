/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';

/**
 * Higher-Order Component to render wrapped components when changed conversion events are detected.
 *
 * @since n.e.x.t
 *
 * @return {Function} Enhancing function.
 */
export default function whenHasChangedConversionEvents() {
	return ( WrappedComponent ) => {
		function WhenHasChangedConversionEvents( props ) {
			const hasChangedConversionEvents = useSelect( ( select ) => {
				const {
					hasNewConversionReportingEvents,
					hasLostConversionReportingEvents,
				} = select( MODULES_ANALYTICS_4 );

				return (
					hasNewConversionReportingEvents() ||
					hasLostConversionReportingEvents()
				);
			}, [] );

			if ( hasChangedConversionEvents ) {
				return <WrappedComponent { ...props } />;
			}

			const { WidgetNull = null } = props;
			return WidgetNull && <WidgetNull />;
		}

		return WhenHasChangedConversionEvents;
	};
}
