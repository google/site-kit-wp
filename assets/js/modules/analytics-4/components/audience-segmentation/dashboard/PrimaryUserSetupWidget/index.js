/**
 * PrimaryUserSetupWidget component.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { useMount } from 'react-use';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import AudienceTileLoading from '@/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/AudienceTile/AudienceTileLoading';
import { MODULES_ANALYTICS_4 } from '@/js/modules/analytics-4/datastore/constants';
import AudienceSegmentationErrorWidget from '@/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSegmentationErrorWidget';
import { isInsufficientPermissionsError } from '@/js/util/errors';
import useEnableAudienceGroup from '@/js/modules/analytics-4/hooks/useEnableAudienceGroup';

export default function PrimaryUserSetupWidget( { Widget } ) {
	const { apiErrors, isSaving, failedAudiences, onEnableGroups } =
		useEnableAudienceGroup();

	const isSettingUpAudiences = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).isSettingUpAudiences()
	);

	useMount( () => {
		if ( isSettingUpAudiences ) {
			return;
		}

		onEnableGroups();
	} );

	if ( ( apiErrors.length || failedAudiences.length ) && ! isSaving ) {
		return (
			<AudienceSegmentationErrorWidget
				Widget={ Widget }
				errors={ apiErrors }
				onRetry={ onEnableGroups }
				failedAudiences={ failedAudiences }
				showRetryButton={
					! isInsufficientPermissionsError( apiErrors[ 0 ] )
				}
			/>
		);
	}

	return (
		<Widget className="googlesitekit-widget-audience-tiles" noPadding>
			<div className="googlesitekit-widget-audience-tiles__body">
				<Widget noPadding>
					<AudienceTileLoading />
				</Widget>
				<Widget noPadding>
					<AudienceTileLoading />
				</Widget>
			</div>
		</Widget>
	);
}

PrimaryUserSetupWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};
