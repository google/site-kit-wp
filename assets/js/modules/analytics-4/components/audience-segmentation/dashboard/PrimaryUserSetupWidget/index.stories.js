/**
 * PrimaryUserSetupWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import fetchMock from 'fetch-mock';

/**
 * Internal dependencies.
 */
import PrimaryUserSetupWidget from '.';
import WithRegistrySetup from '../../../../../../../../tests/js/WithRegistrySetup';
import {
	freezeFetch,
	provideModuleRegistrations,
	provideModules,
	provideUserAuthentication,
} from '../../../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '@/js/googlesitekit/widgets/util';
import { CORE_USER } from '@/js/googlesitekit/datastore/user/constants';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '@/js/util/errors';
import {
	EDIT_SCOPE,
	MODULES_ANALYTICS_4,
} from '@/js/modules/analytics-4/datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '@/js/modules/analytics-4/constants';
import { availableAudiences } from '@/js/modules/analytics-4/datastore/__fixtures__';

const userAuthenticationEndpoint = new RegExp(
	'^/google-site-kit/v1/core/user/data/authentication'
);
const syncAvailableAudiencesEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/sync-audiences'
);

const createAudienceEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/create-audience'
);

const syncAvailableCustomDimensionsEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/sync-custom-dimensions'
);

const WidgetWithComponentProps = withWidgetComponentProps(
	'analyticsAudiencePrimaryUserSetup'
)( PrimaryUserSetupWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const SetupInProgress = Template.bind( {} );
SetupInProgress.storyName = 'Setup in progress';
SetupInProgress.args = {
	setupRegistry: () => {
		freezeFetch( userAuthenticationEndpoint );
	},
};

export const InsufficientPermissionError = Template.bind( {} );
InsufficientPermissionError.storyName = 'Insufficient permission error';
InsufficientPermissionError.args = {
	setupRegistry: ( registry ) => {
		provideUserAuthentication( registry );

		const errorResponse = {
			code: 'test_error',
			message: 'Error message.',
			data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
		};

		fetchMock.post( syncAvailableAudiencesEndpoint, {
			body: errorResponse,
			status: 500,
		} );
	},
};

export const SetupError = Template.bind( {} );
SetupError.storyName = 'Setup error';
SetupError.args = {
	setupRegistry: ( registry ) => {
		provideUserAuthentication( registry );

		const errorResponse = {
			code: 'test_error',
			message: 'Error message.',
			data: { status: 500 },
		};

		fetchMock.post( syncAvailableAudiencesEndpoint, {
			body: errorResponse,
			status: 500,
		} );
	},
};

export const FailedAudience = Template.bind( {} );
FailedAudience.storyName = 'Failed audience';
FailedAudience.args = {
	setupRegistry: ( registry ) => {
		provideUserAuthentication( registry, {
			grantedScopes: EDIT_SCOPE,
		} );
		fetchMock.postOnce( syncAvailableAudiencesEndpoint, {
			body: availableAudiences.slice( 0, 2 ),
			status: 200,
		} );

		fetchMock.postOnce( syncAvailableCustomDimensionsEndpoint, {
			body: [ 'googlesitekit_post_author' ],
			status: 200,
		} );

		const errorResponse = {
			code: 'test_error',
			message: 'Error message.',
			data: { status: 500 },
		};

		fetchMock.postOnce( createAudienceEndpoint, {
			body: errorResponse,
			status: 500,
		} );
	},
};

export default {
	title: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/PrimaryUserSetupWidget',
	decorators: [
		( Story, { args } ) => {
			async function setupRegistry( registry ) {
				provideModules( registry, [
					{
						active: true,
						connected: true,
						slug: MODULE_SLUG_ANALYTICS_4,
					},
				] );
				provideModuleRegistrations( registry );

				const [
					accountID,
					propertyID,
					measurementID,
					webDataStreamID,
				] = [ '12345', '34567', '56789', '78901' ];

				await registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setAccountID( accountID );
				await registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setPropertyID( propertyID );
				await registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setMeasurementID( measurementID );
				await registry
					.dispatch( MODULES_ANALYTICS_4 )
					.setWebDataStreamID( webDataStreamID );

				const audienceSettings = {
					configuredAudiences: null,
					isAudienceSegmentationWidgetHidden: false,
				};

				await registry
					.dispatch( CORE_USER )
					.receiveGetUserAudienceSettings( audienceSettings );

				await args?.setupRegistry( registry );
			}
			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
