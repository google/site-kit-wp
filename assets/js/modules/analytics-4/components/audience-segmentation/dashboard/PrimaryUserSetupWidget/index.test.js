/**
 * PrimaryUserSetupWidget component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import PrimaryUserSetupWidget from '.';
import { render, waitFor } from '../../../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideModules,
	freezeFetch,
	provideUserAuthentication,
	provideModuleRegistrations,
} from '../../../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '@/js/googlesitekit/widgets/util';
import { CORE_USER } from '@/js/googlesitekit/datastore/user/constants';
import {
	EDIT_SCOPE,
	MODULES_ANALYTICS_4,
} from '@/js/modules/analytics-4/datastore/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '@/js/modules/analytics-4/constants';
import { availableAudiences } from '@/js/modules/analytics-4/datastore/__fixtures__';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '@/js/util/errors';

const syncAvailableCustomDimensionsEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/sync-custom-dimensions'
);

const syncAvailableAudiencesEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/sync-audiences'
);

const createAudienceEndpoint = new RegExp(
	'^/google-site-kit/v1/modules/analytics-4/data/create-audience'
);

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsAddMetric'
)( PrimaryUserSetupWidget );

describe( 'PrimaryUserSetupWidget', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();

		provideUserAuthentication( registry, {
			grantedScopes: EDIT_SCOPE,
		} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableCustomDimensions: [],
		} );

		const audienceSettings = {
			configuredAudiences: [],
			isAudienceSegmentationWidgetHidden: false,
		};

		registry
			.dispatch( CORE_USER )
			.receiveGetUserAudienceSettings( audienceSettings );

		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: MODULE_SLUG_ANALYTICS_4,
			},
		] );
		provideModuleRegistrations( registry );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'should start audiences creation on mount', async () => {
		freezeFetch( syncAvailableAudiencesEndpoint );

		const { container } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		await waitFor( () =>
			expect(
				container.querySelectorAll(
					'.googlesitekit-audience-segmentation-tile-loading'
				)
			).toHaveLength( 2 )
		);

		await waitFor( () =>
			expect( fetchMock ).toHaveFetched( syncAvailableAudiencesEndpoint )
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'should handle generic errors and diplay a retry button', async () => {
		fetchMock.post( syncAvailableAudiencesEndpoint, {
			body: {
				code: 'test_error',
				message: 'Error message.',
			},
			status: 500,
		} );

		const { container, getByText } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		await waitFor( () => {
			expect(
				getByText( 'Your visitor groups data loading failed' )
			).toBeInTheDocument();
			expect( getByText( 'Retry' ) ).toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );

	it( 'should display the audiences that failed to be created', async () => {
		fetchMock.post( syncAvailableCustomDimensionsEndpoint, {
			body: [],
			status: 200,
		} );

		fetchMock.post( syncAvailableAudiencesEndpoint, {
			body: availableAudiences.slice( 0, 2 ),
			status: 200,
		} );

		fetchMock.post( createAudienceEndpoint, {
			body: {
				code: 'test_error',
				message: 'Error message.',
				data: { status: 500 },
			},
			status: 500,
		} );

		const { container, getByText } = render( <WidgetWithComponentProps />, {
			registry,
		} );

		await waitFor( () => {
			expect(
				getByText( 'Failed to create the following audiences:' )
			).toBeInTheDocument();
			expect( getByText( 'Retry' ) ).toBeInTheDocument();
			expect( getByText( 'new-visitors' ) ).toBeInTheDocument();
			expect( getByText( 'returning-visitors' ) ).toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );

	it( 'should handle the insufficient permissions error', async () => {
		fetchMock.post( syncAvailableAudiencesEndpoint, {
			body: {
				code: 'test_error',
				message: 'Error message.',
				data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
			},
			status: 403,
		} );

		const { container, queryByText, getByText } = render(
			<WidgetWithComponentProps />,
			{
				registry,
			}
		);

		await waitFor( () => {
			expect(
				getByText( 'Insufficient permissions' )
			).toBeInTheDocument();
			expect( queryByText( 'Retry' ) ).not.toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );
} );
