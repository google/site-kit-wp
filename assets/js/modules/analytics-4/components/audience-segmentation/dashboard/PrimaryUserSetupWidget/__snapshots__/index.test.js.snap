// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryUserSetupWidget should display the audiences that failed to be created 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-typography googlesitekit-typography--title googlesitekit-typography--large"
            >
              Your visitor groups data loading failed
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <div>
                <p
                  class="googlesitekit-typography googlesitekit-typography--body googlesitekit-typography--medium"
                >
                  Failed to create the following audiences:
                </p>
                <ul
                  class="googlesitekit-widget-audience-segmentation-error__failed-audiences"
                >
                  <li>
                    new-visitors
                  </li>
                  <li>
                    returning-visitors
                  </li>
                </ul>
                <button
                  class="mdc-button mdc-button--raised mdc-button--danger"
                  target="_self"
                >
                  <span
                    class="mdc-button__label"
                  >
                    Retry
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell mdc-layout-grid__cell--span-5-desktop mdc-layout-grid__cell--span-3-tablet"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PrimaryUserSetupWidget should handle generic errors and display a retry button 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-typography googlesitekit-typography--title googlesitekit-typography--large"
            >
              Your visitor groups data loading failed
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <button
                class="mdc-button mdc-button--raised mdc-button--danger"
                target="_self"
              >
                <span
                  class="mdc-button__label"
                >
                  Retry
                </span>
              </button>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell mdc-layout-grid__cell--span-5-desktop mdc-layout-grid__cell--span-3-tablet"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PrimaryUserSetupWidget should handle the insufficient permissions error 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding googlesitekit-audience-segmentation-error-widget"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="mdc-layout-grid googlesitekit-audience-segmentation-error__widget-primary-cell mdc-layout-grid--collapsed"
      >
        <div
          class="mdc-layout-grid__inner"
        >
          <div
            class="mdc-layout-grid__cell mdc-layout-grid__cell--span-7-desktop mdc-layout-grid__cell--span-8-tablet"
          >
            <h3
              class="googlesitekit-typography googlesitekit-typography--title googlesitekit-typography--large"
            >
              Insufficient permissions
            </h3>
            <div
              class="googlesitekit-widget-audience-segmentation-error__actions"
            >
              <div
                class="googlesitekit-report-error-actions"
              >
                <div
                  class="googlesitekit-error-retry-text"
                >
                  Contact your administrator. Trouble getting access? 
                  <a
                    aria-label="Get help (opens in a new tab)"
                    class="googlesitekit-cta-link"
                    href="undefined?error_id=analytics-4_insufficient_permissions"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    <span
                      class="googlesitekit-cta-link__contents"
                    >
                      Get help
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="googlesitekit-widget-audience-segmentation-error__svg-wrapper mdc-layout-grid__cell mdc-layout-grid__cell--span-5-desktop mdc-layout-grid__cell--span-3-tablet"
          >
            <svg />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PrimaryUserSetupWidget should start audiences creation on mount 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding googlesitekit-widget-audience-tiles"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-widget-audience-tiles__body"
      >
        <div
          class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding"
        >
          <div
            class="googlesitekit-widget__body"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-loading"
            >
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 20px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="googlesitekit-widget googlesitekit-widget--keyMetricsAddMetric googlesitekit-widget--no-padding"
        >
          <div
            class="googlesitekit-widget__body"
          >
            <div
              class="googlesitekit-audience-segmentation-tile-loading"
            >
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 20px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
              <div
                class="googlesitekit-preview-block"
                style="width: 100%; height: 52px;"
              >
                <div
                  class="googlesitekit-preview-block__wrapper"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
